File: src/time_based_key_value_store_981/solution.go
package time_based_key_value_store_981

import "fmt"

type ValueAtTime struct {
	value     string
	timestamp int
}

type TimeMap struct {
	store map[string][]ValueAtTime
}

func Constructor() TimeMap {
	return TimeMap{
		store: make(map[string][]ValueAtTime),
	}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	newValueAtTime := ValueAtTime{value: value, timestamp: timestamp}
	this.store[key] = append(this.store[key], newValueAtTime)
}

func (this *TimeMap) Get(key string, timestamp int) string {
	values, found := this.store[key]
	if !found {
		return ""
	}

	l := 0
	r := len(values)

	for l < r {
		mid := (l + r) / 2

		if values[mid].timestamp <= timestamp {
			l = mid + 1
		} else {
			r = mid
		}
	}

	if r == 0 {
		return ""
	}

	return values[r-1].value
}

// Test
/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */
func Test() {
	obj := Constructor()
	obj.Set("foo", "bar", 1)
	fmt.Println(obj.Get("foo", 1))
	fmt.Println(obj.Get("foo", 3))
	obj.Set("foo", "bar2", 4)
	fmt.Println(obj.Get("foo", 4))
	fmt.Println(obj.Get("foo", 5))
}

File: src/time_based_key_value_store_981/solution.rs
use std::collections::HashMap;

#[derive(Clone)]
struct ValueAtTime {
    value: String,
    timestamp: i32,
}

struct TimeMap {
    store: HashMap<String, Vec<ValueAtTime>>,
}

impl TimeMap {
    fn new() -> Self {
        TimeMap {
            store: HashMap::new(),
        }
    }

    fn set(&mut self, key: String, value: String, timestamp: i32) {
        let entry = self.store.entry(key).or_default();
        entry.push(ValueAtTime { value, timestamp });
    }

    fn get(&self, key: String, timestamp: i32) -> String {
        let Some(values) = self.store.get(&key) else {
            return "".to_string();
        };

        let mut left = 0;
        let mut right = values.len();

        while left < right {
            let mid = (left + right) / 2;
            if values[mid].timestamp <= timestamp {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        if right == 0 {
            return "".to_string();
        }

        values[right - 1].value.clone()
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * let obj = TimeMap::new();
 * obj.set(key, value, timestamp);
 * let ret_2: String = obj.get(key, timestamp);
 */
pub fn test() {
    let mut obj = TimeMap::new();
    obj.set("foo".to_string(), "bar".to_string(), 1);
    println!("{}", obj.get("foo".to_string(), 1));
    println!("{}", obj.get("foo".to_string(), 3));
    obj.set("foo".to_string(), "bar2".to_string(), 4);
    println!("{}", obj.get("foo".to_string(), 4));
    println!("{}", obj.get("foo".to_string(), 5));
}
